compiler/rustc_mir_transform/src/coverage/mod.rs

```diff
@@ -189,27 +198,94 @@ fn create_mappings<'tcx>(
             Some(Mapping { kind, code_region })
         },
     ));
+    let mut term_for_bcbs = |bcbs: &[BasicCoverageBlock]| {
+        let counters = bcbs
+            .into_iter()
+            .copied()
+            .map(|bcb| {
+                coverage_counters.bcb_counter(bcb).expect("all BCBs with spans were given counters")
+            })
+            .collect::<Vec<_>>();
+        let counter = counters
+            .into_iter()
+            .fold(None, |acc, val| Some(coverage_counters.make_sum_expression(acc, val)))
+            .expect("counters must be non-empty");
+        // Since the counter is a sum of multiple counters, it should be marked used in
+        // all associated bcbs in case it were markered as `Zero`.
+        if bcbs.len() > 1
+            && let counters::BcbCounter::Expression { id } = counter
+        {
+            for &bcb in bcbs {
+                inject_statement(
+                    mir_body,
+                    CoverageKind::ExpressionUsed { id },
+                    basic_coverage_blocks[bcb].leader_bb(),
+                );
+            }
+        }
```

Comment:
Hmm, I'm not sure about this.

Is the idea that the sum should be simplified to `Zero` if all of its terms are zero? If so, it seems like it should be possible to detect that situation during codegen, without having to inject extra `ExpressionUsed` statements for the sum expression.

Reply:
If this expression was not marked with these `ExpressionUsed` statements in associated blocks, it can not be called with `FunctionCoverageCollector::mark_expression_id_seen`. Then compiler will take it as unused and simplify it with `Zero` directly. This code can make sure it is simplified with `Zero` only when all of the bcbs are unreachable.

Reply:
I think the proper solution is to change the `ExpressionUsed` code in codegen to only care about `MappingKind::Code`, and ignore any expressions used by branch or MC/DC mappings.

I'll prepare a separate PR to make that change.

Reply:
#127758

Reply:
Excellent, that works for pattern matching. I've removed these code at [47ff2f5](https://github.com/rust-lang/rust/pull/126677/commits/47ff2f5950abf75efc7f67dce755f68562bc0542)